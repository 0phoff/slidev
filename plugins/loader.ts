import { resolve } from 'path'
import { Plugin } from 'vite'
import fs from 'fs-extra'
import { isNotNull } from '@antfu/utils'
import matter from 'gray-matter'

const filepath = resolve(__dirname, '../slides.md')

async function read() {
  return await fs.readFile(resolve(__dirname, '../slides.md'), 'utf-8')
}

function parse(raw: string) {
  const lines = raw.split(/\n/g)
  const pages: string[] = []
  let start = 0
  let dividers = 0

  lines.forEach((line, i) => {
    if (line === '---')
      dividers += 1

    if (dividers === 3) {
      pages.push(lines.slice(start, i).join('\n'))
      dividers = 1
      start = i
    }
  })

  if (start !== lines.length - 1)
    pages.push(lines.slice(start).join('\n'))

  return pages
}

export function createSlidesLoader(): Plugin {
  let raw: string | undefined
  let items: string[] = []

  return {
    name: 'slides-loader',

    async configResolved() {
      raw = await read()
      items = parse(raw)
    },

    async handleHotUpdate(ctx) {
      if (ctx.file === filepath) {
        raw = await ctx.read()
        items = parse(raw)

        return items
          .map((i, idx) => ctx.server.moduleGraph.getModuleById(`/@vite-slides/slides/${idx}.md`))
          .filter(isNotNull)
      }
    },

    resolveId(id) {
      if (id.startsWith('/@vite-slides/'))
        return id
      return null
    },

    load(id) {
      const match = id.match(/^\/\@vite-slides\/slide\/(\d+).md$/)
      if (match) {
        const pageNo = parseInt(match[1])
        return items[pageNo]
      }
      else if (id === '/@vite-slides/routes') {
        const imports: string[] = []

        const routes = `export default [\n${
          items
            .map((i, idx) => {
              imports.push(`import n${idx} from '/@vite-slides/slide/${idx}.md'`)
              const { data: meta } = matter(i)
              return `{ path: '/${idx}', name: 'page-${idx}', component: n${idx}, meta: ${JSON.stringify(meta)} },\n`
            })
            .join('')
        }]`

        return [...imports, routes].join('\n')
      }
    },
  }
}
