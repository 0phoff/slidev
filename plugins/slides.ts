import { resolve } from 'path'
import { Plugin } from 'vite'
import fs from 'fs-extra'
import { notNullish } from '@antfu/utils'
import matter from 'gray-matter'

const filepath = resolve(__dirname, '../slides.md')

async function read() {
  return await fs.readFile(resolve(__dirname, '../slides.md'), 'utf-8')
}

function parse(raw: string) {
  const lines = raw.split(/\n/g)
  const pages: string[] = []
  let start = 0
  let dividers = 0

  lines.forEach((line, i) => {
    line = line.trimRight()

    if (line === '---')
      dividers += 1

    const isHardDivider = line === '------'

    if (dividers >= 3 || isHardDivider) {
      pages.push(lines.slice(start, isHardDivider ? i - 1 : i).join('\n'))
      dividers = isHardDivider ? 2 : 1
      start = isHardDivider ? i + 1 : i
    }
  })

  if (start !== lines.length - 1)
    pages.push(lines.slice(start).join('\n'))

  pages.push('---\nlayout: end\n---')

  return pages
}

export function createSlidesLoader(): Plugin {
  let raw: string | undefined
  let items: string[] = []

  return {
    name: 'vite-slides:loader',

    async configResolved() {
      raw = await read()
      items = parse(raw)
    },

    configureServer(server) {
      server.watcher.add(filepath)
    },

    async handleHotUpdate(ctx) {
      if (ctx.file === filepath) {
        raw = await read()
        items = parse(raw)

        const moduleEntries = [
          '/@vite-slides/routes',
          ...items.map((i, idx) => `/@vite-slides/slide/${idx}.md`),
        ]
          .map(id => ctx.server.moduleGraph.getModuleById(id))
          .filter(notNullish)

        moduleEntries.map(m => ctx.server.moduleGraph.invalidateModule(m))
        return moduleEntries
      }
    },

    resolveId(id) {
      if (id.startsWith('/@vite-slides/'))
        return id
      return null
    },

    load(id) {
      const match = id.match(/^\/\@vite-slides\/slide\/(\d+).md$/)
      if (match) {
        const pageNo = parseInt(match[1])
        return items[pageNo]
      }
      else if (id === '/@vite-slides/routes') {
        const imports: string[] = []

        const routes = `export default [\n${
          items
            .map((i, idx) => {
              imports.push(`import n${idx} from '/@vite-slides/slide/${idx}.md'`)
              const { data: meta } = matter(i)
              return `{ path: '/${idx}', name: 'page-${idx}', component: n${idx}, meta: ${JSON.stringify(meta)} },\n`
            })
            .join('')
        }]`

        return [...imports, routes].join('\n')
      }
    },
  }
}
