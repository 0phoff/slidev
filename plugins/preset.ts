
import { mergeConfig, Plugin } from 'vite'
import Vue from '@vitejs/plugin-vue'
import ViteIcons, { ViteIconsResolver } from 'vite-plugin-icons'
import ViteComponents from 'vite-plugin-components'
import Markdown from 'vite-plugin-md'
import WindiCSS from 'vite-plugin-windicss'
import Prism from 'markdown-it-prism'
import base64 from 'js-base64'
import { createSlidesLoader } from './slides'
import { createMonacoLoader } from './monaco'

export type ArgumentsType<T> = T extends ((...args: infer A) => void) ? A : never

export interface ViteSlidesOptions {
  vue?: ArgumentsType<typeof Vue>[0]
  markdown?: ArgumentsType<typeof Markdown>[0]
  components?: ArgumentsType<typeof ViteComponents>[0]
  windicss?: ArgumentsType<typeof WindiCSS>[0]
  icons?: ArgumentsType<typeof ViteIcons>[0]
}

export function ViteSlides(options: ViteSlidesOptions = {}): Plugin[] {
  const {
    vue: vueOptions = {},
    markdown: mdOptions = {},
    components: componentsOptions = {},
    windicss: windicssOptions = {},
    icons: iconsOptions = {},
  } = options

  return [
    {
      name: 'vite-slides:config',
      config(config) {
        return mergeConfig(config, {
          optimizeDeps: {
            include: [
              'vue',
              'vue-router',
              '@vueuse/core',
            ],
            exclude: [
              'vue-demi',
            ],
          },
        })
      },
    },

    Vue({
      include: [/\.vue$/, /\.md$/],
      ...vueOptions,
    }),

    Markdown({
      wrapperClasses: '',
      headEnabled: true,
      markdownItOptions: {
        quotes: '""\'\'',
        html: true,
        xhtmlOut: true,
        linkify: true,
      },
      markdownItSetup(md) {
      // https://prismjs.com/
        md.use(Prism)
      },
      transforms: {
        before(code) {
        // transform monaco
          code = code.replace(/\n```(\w+?){monaco([\w:,-]*)}[\s\n]*([\s\S]+?)\n```/mg, (full, lang = 'ts', options: string, code: string) => {
            options = options || ''
            code = base64.encode(code, true)
            return `<Monaco :code="'${code}'" :lang="'${lang}'" :readonly="${options.includes('readonly')}" />`
          })

          return code
        },
      },
      ...mdOptions,
    }),

    {
      name: 'vite-slides:vue-escape',
      enforce: 'post',
      transform(code, id) {
        if (id.endsWith('.md'))
          return code.replace(/\\{/g, '{')
      },
    },

    ViteComponents({
      extensions: ['vue', 'md', 'ts'],

      dirs: [
        'src/builtin',
        'src/components',
      ],

      customLoaderMatcher: id => id.endsWith('.md'),

      customComponentResolvers: [
        ViteIconsResolver({
          componentPrefix: '',
        }),
      ],

      ...componentsOptions,
    }),

    createSlidesLoader(),
    createMonacoLoader(),

    ViteIcons({
      ...iconsOptions,
    }),

    ...WindiCSS({
      ...windicssOptions,
    }),
  ]
}
