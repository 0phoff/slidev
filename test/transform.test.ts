import { transformSlotSugar } from '../packages/slidev/node/plugins/markdown'

describe('markdown transform', () => {
  it('slot-sugar', () => {
    expect(transformSlotSugar(`
# Page 

Default Slot
::right::
Right Slot
::left::
<div>Left Slot</div>
`)).toMatchSnapshot()
  })

  it('slot-sugar with code', () => {
    expect(transformSlotSugar(`
# Page 

Default Slot

::code::

\`\`\`md
Slot Usage
::right::
::left::
\`\`\`

`)).toMatchSnapshot()
  })
})
