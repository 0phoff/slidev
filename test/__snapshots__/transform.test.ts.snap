// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markdown transform slot-sugar 1`] = `
"
# Page 

Default Slot

<template v-slot:right>

Right Slot

</template>
<template v-slot:left>

<div>Left Slot</div>

</template>"
`;

exports[`markdown transform slot-sugar with code 1`] = `
"
# Page 

Default Slot


<template v-slot:code>


\`\`\`md
Slot Usage
::right::
::left::
\`\`\`


</template>"
`;
